coordinator/src/main/scala/filodb.coordinator/IngestionActor.scala

```diff
@@ -339,21 +349,14 @@ private[filodb] final class IngestionActor(dataset: Dataset,
     origin ! IngestionStatus(memStore.numRowsIngested(dataset.ref))
 
   private def stopIngestion(shard: Int): Unit = {
+    // When the future is canceled, the onCancel task installed earlier should run.
     logger.info(s"stopIngestion called for dataset=${dataset.ref} shard=$shard")
-    val shardIngestion = streamSubscriptions.get(shard)
-    shardIngestion.foreach { s =>
-      s.onComplete {
-        case Success(_) =>
-          // release resources when stop is invoked explicitly, not when ingestion ends in non-kafka environments
-          removeAndReleaseResources(dataset.ref, shard)
-          logger.info(s"stopIngestion handler done. Ingestion success for dataset=${dataset.ref} shard=$shard")
-          // ingestion stopped event is already handled in the normalIngestion method
-        case Failure(f) =>
-          logger.error(s"stopIngestion handler done. Ingestion failure for dataset=${dataset.ref} shard=$shard", f)
-          // release of resources on failure is already handled in the normalIngestion method
-      }(actorDispatcher)
-    }
-    shardIngestion.foreach(_.cancel())
+    streamSubscriptions.get(shard).foreach(_.cancel())
+  }
+
+  private def ingestionStopped(ref: DatasetRef, shard: Int): Unit = {
+    removeAndReleaseResources(ref, shard)
```

Comment:
This could be done on the endHandler itself. This way it would get scheduled on the ingestion thread.

Reply:
This creates ordering races with the ingestion actor.